diff --git a/node_modules/blockly/blockly.d.ts b/node_modules/blockly/blockly.d.ts
index 291f472..f4a923b 100644
--- a/node_modules/blockly/blockly.d.ts
+++ b/node_modules/blockly/blockly.d.ts
@@ -18942,6 +18942,28 @@ declare module "events/events_block_delete" {
 }
 declare module "block" {
     export class Block {
+        
+        /**
+         * Added by spellsource. Stores the json representation of this block.
+         */
+        json?: any
+        /**
+         * Added by spellsource. Used by art blocks to store whether they've been used on cards or not.
+         */
+        used?: boolean
+        /**
+         * Added by spellsource.
+         */
+        message?: string
+        /**
+         * Added by spellsource. The CardScript representation of this block, if it's a card
+         */
+        cardScript?: any;
+        /**
+         * Added by spellsource. The generated card id for this block, if it's a card
+         */
+        cardId: string
+
         /**
          * Class for one block.
          * Not normally called directly, workspace.newBlock() is preferred.
